README.md
mojolicious-plugin-mozpersona
=============================

Minimalistic integration of Mozillas "Persona" authentication system in Mojolicious apps

Mojolicious-Plugin-MozPersona/Changes

{{$NEXT}}

0.05      2015-01-25 19:52:23+01:00 Europe/Berlin
	- Deprecate the default meaning of the "autoHook"/"jquery"
	  option to automatically include the jQuery file that
	  comes bundled with Mojolicious: it is intended for
	  internal use only (kraih)
	- Use functional Mojo::JSON API since the object-oriented
	  API was removed in Mojolicious 5.73; require at least 
	  Mojolicious v4.82, since this was when decode_json was added
	  Patch provided by THUNTER

0.04      2014-01-02 19:34:05 Europe/Berlin
	- Fix for newer Mojo: use UA->post($url, form => {...})
	  instead of the now deprecated UA->post_form
	  Patch by moritz - thanks!
	- Use default URL of jQuery library which is part of the
	  Mojolicious distribution
	  Patch by moritz - thanks!

0.03      2013-01-27 18:36:07 Europe/Berlin
	- Forgot to explicitly mention Mojolicious as dependency
	  as CPAN testers reminded me

0.02      2013-01-24 22:59:34 Europe/Berlin
	- Minor doc changes
	- Moved to github

0.01      2013-01-23 23:13:21 Europe/Berlin
	- Only changed the version number to have the dist
	  indexed by CPAN

0.01_01   2013-01-23 21:28:37 Europe/Berlin
	Initial release


Mojolicious-Plugin-MozPersona/lib/Mojolicious/Plugin/MozPersona/Controller.pm
use strict;
use warnings;

package Mojolicious::Plugin::MozPersona::Controller;

# ABSTRACT: Default implementation for server side functions for "Persona" authentication.

use Mojo::Base 'Mojolicious::Controller';

use Mojo::JSON qw(decode_json);

use Mozilla::CA qw();
use Data::Dumper;

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::MozPersona::Controller - Default implementation for server side functions for "Persona" authentication.

B<BEWARE: THIS IS AN ALPHA RELEASE!> - It is quite possible that there will
be incompatible changes before the API and the functionality of this 
plugin are declared stable. Use at your own risk.
On the other hand that also means that now is the right time to tell me
what should be changed to make this module more usable and useful!

=head1 DESCRIPTION

When registering the "MozPersona" plugin you may specify a configuration
param C<namespace> which must contain the name of a module that inherits
from L<Mojolicious::Controller>.

This module must provide (directly or via inheritance) implementations
of three methods: C<signin>, C<signout> and C<js>.

=head1 METHODS

=head2 signin

This method gets called via an XMLHttpRequest after a user signed in with
the identity provider - i.e. after the IdP created and stored an association
between the user and the given audience (the Mojolicious app).

It receives a controller object which has access to the authentication 
assertion of the IdP as request param C<assertion>. 
It also has access to the configuration values "audience" (as 
C<_persona_audience>) and "service" (as C<_persona_service>) via the stash.

It is responsible for sending a request containing the assertion and the
audience to the service, thereby verifying the assertion token and 
retrieves a JSON response which tells if the assertion was verified,
until when it is valid and what email address was associated with the user.

It should update the Mojolicious session accordingly. The default
implementation in this module does so by saving the IdP response in the
C<_persona> key in the session). 

Please note that the session is updated with the IdP response no matter if
the authentication and verification actually succeeded! Simply checking
for the existence of C<session->{_persona}> is B<NOT> sufficient - you need
to check the value of the C<status> element in the response of the Persona IdP.
Setting the "_persona" data anyway allows the app to provide a failure 
message for subsequent requests by the user.

This method should return a status info to the browser: The default implementation
in this module returns C<"success":false> if there was an error while
processing the request and C<"success":true> otherwise.

C<"success":true>, however, does B<NOT> indicate that the user was successfully
authenticated! Again, you have to refer to the message returned by the IdP which is also
included in the response and contains either C<"status":"okay"> or C<"status":"failure">.

E.g. if the user successfully signed in and the verification succeeded the JSON
message returned is:

  {
    "success": true,
    "result": {
      "status"  : "okay",
      "email"   : "somebody@example.org",
      "audience": "http:\/\/127.0.0.1:3000\/",
      "issuer"  : "login.persona.org",
      "expires" : 1358864866835
    }
  }

Otherwise if the message exchange was successful but IdP could not verify the
authentication of the user the JSON message looks like this:

  {
    "success": true,
    "result": {
      "status": "failure",
      "reason": "...server hiccup..."
    }
  }

And if the whole process failed the JSON message simply is

  {
    "success": false
  }

For further info please refer to L<https://developer.mozilla.org/en-US/docs/Persona>, 
L<https://developer.mozilla.org/en-US/docs/Persona/Quick_Setup> and 
L<https://developer.mozilla.org/en-US/docs/Persona/Remote_Verification_API>.

=cut

sub signin {
    my $self = shift;

    $ENV{'MOJO_CA_FILE'} = Mozilla::CA::SSL_ca_file();

    my $persona_response = '';
    my $result = '';

    eval {
        $persona_response = $self->ua->post(
           $self->stash('_persona_service')
           => form => {
                assertion => $self->param('assertion'),
                audience  => $self->stash('_persona_audience'), 
           }
        )->res;

        $result = decode_json $persona_response->body;
    };

    if ($@) {
        $self->app->log->error("Error verifying assertion with IdP: $@");
        $self->render( json => { signin => Mojo::JSON->false } );
    }
    elsif ( ! ( $result->{'status'} eq "okay" or $result->{'status'} eq "failure" ) ) {
        require Data::Dumper;
        $self->app->log->error("Invalid response from IdP: " . Data::Dumper::Dumper($result));
    }
    else {
        if ( $self->app->log->is_debug ) {
            require Data::Dumper;
            $self->app->log->debug("Successfully verified user assertion with IdP: " . Data::Dumper::Dumper($result));
        }
        $self->session->{_persona} = $result;
        $self->render( json => { signin => Mojo::JSON->true, result => $result } );
    }
}

=head2 signout

This method gets called via an XMLHttpRequest after a user successfully
signed out with the identity provider.

It receives a controller object which has access to the Mojolicious session.

It should remove the authentication information from the sessions. The
default implementation in this module does so by simply deleting the key
"_persona" from the session hash.

It should return a status info to the browser. Since all that this default
implementation of the method does is delete the Persona info from the 
session it always returns C<{"signout":true}>.

=cut

sub signout {
    my $self = shift;
    delete $self->session->{_persona};
    $self->render( json => { signout => Mojo::JSON->true } );
}

=head2 js

This method gets called from the browser to provide a JavaScript document
that registers C<click> handlers for the login and logout button. The
JavaScript code also registers the appropriate callback functions with
the C<watch> method of the browsers C<navigator.id> API.
See also L<https://developer.mozilla.org/en-US/docs/DOM/navigator.id>.

The method retrieves the complete config of the Mojolicious::Plugin::MozPersona
module as a hash in the stash slot C<_persona_conf>.

Using this info it renders the template specified as C<localJsTpl> in
the configuration.

=cut

sub js {
    my $self = shift;
    my %c = %{ $self->stash('_persona_conf') };

    foreach my $k ( keys %c ) {
        $self->stash( $k => $c{$k} );
    }

    # set empty value for optional config
    foreach my $w ( qw( siteLogo privacyPolicy termsOfService returnTo oncancel ) ) {
        $self->stash( $w => '' ) unless $c{$w};
    }

    $self->res->headers->content_type('text/javascript');

    my ( $tName, $tFormat, $tHandler ) = split( /\./, $c{'localJsTpl'} );
    $tName    = 'persona_local_js' unless defined($tName)    and $tName;
    $tFormat  = 'txt'              unless defined($tFormat)  and $tFormat;
    $tHandler = 'ep'               unless defined($tHandler) and $tHandler;

    $self->render( $tName, format => $tFormat, handler => $tHandler );
}

=head1 SEE ALSO

L<Mojolicious::Plugin::MozPersona>,
L<https://developer.mozilla.org/en-US/docs/Persona>,
L<Mojolicious>.

=cut

1;

__END__



Mojolicious-Plugin-MozPersona/lib/Mojolicious/Plugin/MozPersona.pm

package Mojolicious::Plugin::MozPersona;

# ABSTRACT: Minimalistic integration of Mozillas "Persona" authentication system in Mojolicious apps

use strict;
use warnings;

=pod

=encoding utf8

=head1 NAME

Mojolicious::Plugin::MozPersona - Minimalistic integration of Mozillas 
"Persona" authentication system in Mojolicious apps.

B<BEWARE: THIS IS AN EARLY RELEASE!> - It is quite possible that there will
be incompatible changes before the API and the functionality of this 
plugin are declared stable. Use at your own risk.
On the other hand that also means that now is the right time to tell me
what should be changed to make this module more usable and useful!

=cut

use Mojo::Base 'Mojolicious::Plugin';

use File::Basename 'dirname';
use File::Spec::Functions 'catdir';

use Mojolicious::Plugin::MozPersona::Controller;

=head1 SYNOPSIS

  # Mojolicious::Lite
  plugin 'MozPersona' => {
      audience => 'http://127.0.0.1:3000/',
      siteName => 'My spectacular new site!',
      autoHook => { css => 1 },
  };

  # Mojolicious
  # $self->plugin( moz_persona => { 
  #     audience => 'https://example.org:8443/',
  #     siteName => 'My shiny new site!'
  # });

  get '/' => sub {
      my $self = shift;
      $self->render('index');
  };

  app->start;

  __DATA__

  @@ index.html.ep
  <!DOCTYPE html>
  <html>
    <head><title>Mozilla Persona Test</title></head>
    <body>
  % if ( defined(session '_persona') && (session '_persona')->{'status'} eq 'okay' ) {
  %     my $email = (session '_persona')->{'email'};
      <a href="#" class="persona-button" id="personaSignout"><span><%= $email %> abmelden</span></a>
  % } else {
      <a href="#" class="persona-button" id="personaSignin"><span>Anmelden mit Ihrer Email</span></a>
  % }
      <p>Some dummy content.</p>
    </body>
  </html>:confirm b6


=head1 DESCRIPTION

L<Mojolicious::Plugin::MozPersona> is a L<Mojolicious> plugin.
It provides a minimalistic integration of Mozillas "Persona" authentication
system in Mojolicious apps. 

This modules adds a few routes (at C<signinPath>, C<signoutPath> and C<localJsPath>,
see below) to your app which refer to the code that is responsible for handling
the local, application-specific server side part of the B<Persona> authentication model.

It also by default registers a C<after_dispatch> hook that automatically inserts
some C<script> and C<link> elements to pull in the JavaScript code Persona needs
and some CSS that provides common styling rules for the signin/signout elements
(cf. C<autoHook> below).

Please note that you´re advised to read the
L<Persona documentation|https://developer.mozilla.org/en-US/docs/Persona> before 
using this plugin.

=head1 CONFIGURATION

You may pass a hash ref on plugin registration. The following keys are currently 
recognized: 

=over 4

=item audience

The protocol, domain name, and port of your site. E.g. C<https://example.org:8443>.
This param is mandatory.

=item siteName

The human readable name of your site (the running Mojolicious app).
Will be shown to the user when he or she signs in with the identity 
provider. This param is mandatory.

=item service

The URL of the API of the identity provider you selected.
Default is C<https://verifier.login.persona.org/verify>.

=item namespace

The name of a Mojolicious::Controller module.
This controller provides the server-side functionality necessary to verify
a user login.
Default is L<Mojolicious::Plugin::MozPersona::Controller>.

=item signinId

The id of the HTML element which provides the "onClick" handler to start
the signin process. Default is C<personaSignin>.

=item signinPath

The URL path on your server that get´s called with the assertion of the
identity provider after the user signed in with the identity provider.
The handler code for this URL is responsible to verify the users identity
assertion and update the mojolicious session accordingly. 
Default is C</_persona/signin>.

=item signoutId

The id of the HTML element which provides the "onClick" handler to start
the signout process. Default is C<personaSignout>.

=item signoutPath

The URL path on your server that get´s called when user logs out.
The handler code for this URL is responsible for removing the authentication
information from the Mojolicious session. Default is C</_persona/signout'>.

=item autoHook

Hashref that tells the plugin wether to automatically add html elements to
include necessary JavaScript and CSS components.
Known values are:

=over 4

=item css

Include the style rules for CSS "Buttons" as provided by Mozilla by inserting
a C<E<lt>link rel="stylesheet" href="/_persona/persona-buttons.css"E<gt>> element.
Note that the CSS file is part of this distribution and not loaded from a 
Mozilla server.
Cf. L<https://developer.mozilla.org/en-US/docs/persona/branding>.
This option is deactivated (set to C<0>) by default.

=item jquery

Include the jQuery JavaScript library. The exact effect of activating this option
depends on the value provided:

=over 4

=item C<bundled>

If the option is set to C<bundled>, the jQuery file that is part of the Mojolicious 
distribution is used by inserting a C<E<lt>script src="/mojo/jquery/jquery.js"E<gt>> 
element.

<B>Beware:<B> Since the jQuery library bundled with Mojolicious is intended for internal
use only, this is not a recommended practice: consider using your own independent copy 
or use the CDN.

=item C<cdn>

Include the latest version of the jQuery JavaScript library by
utilizing the official content delivery network; inserts 
C<E<lt>script src="http://code.jquery.com/jquery-latest.min.js"E<gt>>.

As of now, the latest version from the 1.x line of development is available at that
URL. To minimize the risk of future incompatibilities please consider using your own 
independent copy (and set this option to C<false>).

=back

Prior to release 0.05 this option only took a boolean value and if that was C<true> the
bundled jQuery file was used. That is the reason why in release 0.05 any C<true> value 
besides C<bundled> and C<cdn> still has the same effect as if C<bundled> was passed in.

A future release will change this option to be off by default and will consider any value
besides the two shown above as C<false>. 

=item persona

Include the Persona JavaScript code by inserting the appropriate 
C<E<lt>script src="$config{'personaJsPath'}"E<gt>> element. See config param C<personaJs> below.

=item local

Include the local JavaScript code that triggers the start of the login and
logout processes and connects the persona authentication to the session of
the local application by inserting the appropriate C<E<lt>script src="$config{'localJsPath'}"E<gt>>

=item uid

Make the email address of the persona user available to the local JavaScript
code by setting a JavaScript variable.

=back

All hooks except for the C<css> hook are active by default.

Please note that these options are interconnected. E.g. the default local 
JavaScript code B<needs> jQuery so if you keep C<local> active and disable
C<jquery> you must include jQuery yourself! Similarly the default local JS 
code expects a JS variable with a specific name set to either "null" or the
email address of the currenly logged-in user.

=item localJsPath

The URL to use for retrieving the JavaScript code that registers the C<onClick>
handler and callback functions which are necessary for performing the signin
and signout process. Default is C</_persona/localjs>.

=item localJsTpl

The name of the template which provides the locally defined JavaScript "glue"
between your app and the Persona IdP. Default is C<_persona/local_js.txt.ep>.
The given value is automatically split to determine which format and handler
to use for rendering.
Unless specified via the given value C<format =E<gt> text> and C<handler =E<gt> ep>
are used as defaults.

=item personaJsPath

The URL or path to the Persona JavaScript code. 
Default is L<https://login.persona.org/include.js>.

=back

All configuration params the have a default value listed above are optional.

Besides C<siteName> the L<"navigator.id.request" function|https://developer.mozilla.org/en-US/docs/DOM/navigator.id.request>
additionally supports the following params: C<siteLogo>, C<termsOfService>, 
C<privacyPolicy>, C<oncancel>, and C<returnTo>.
You may provide string values for all of these when registering the plugin.
The default local JavaScript code, however, ignores C<oncancel> and C<returnTo>
and only passes the other three params to the Persona IdP.

=cut

my %defaults = (
	audience      => '',
	siteName      => '',
	service       => 'https://verifier.login.persona.org/verify',
	namespace     => 'Mojolicious::Plugin::MozPersona::Controller',
	signinId      => 'personaSignin',
	signinPath    => '/_persona/signin',
	signoutId     => 'personaSignout',
	signoutPath   => '/_persona/signout',
	autoHook      => { css => 0, jquery => 'bundled', persona => 1, local => 1, uid => 1 },
	localJsPath   => '/_persona/localjs',
	localJsTpl    => '_persona/local_js.txt.ep',
	personaJsPath => 'https://login.persona.org/include.js',
);

=head1 METHODS

L<Mojolicious::Plugin::MozPersona> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones:

=head2 C<register>

  $plugin->register(Mojolicious->new);

You usually don´t call that method yourself directly. Instead, Mojolicious 
calls this function when you register the plugin, e.g. via 
C<plugin 'MozPersona' ...>.
See DESCRIPTION above for an explanation of what happens here.

=cut

sub register {
	my ( $self, $app ) = @_;

	my $defaultHooks = delete $defaults{'autoHook'};
	my (%conf) = ( %defaults, %{ $_[2] || {} } );

	$conf{'autoHook'} = {} unless exists $conf{'autoHook'};
	foreach my $h ( keys %{$defaultHooks} ) {
		if ( ref $conf{'autoHook'} && !exists $conf{'autoHook'}->{$h} ) {
			$conf{'autoHook'}->{$h} = $defaultHooks->{$h};
		}
	}

	$conf{'siteName'} =~ tr/"'//d;
	$conf{'signinPath'} =~ tr/"'//d;
	$conf{'signoutPath'} =~ tr/"'//d;
	$conf{'signinId'} =~ tr/"'#//d;
	$conf{'signoutId'} =~ tr/"'#//d;

	die "Missing required configuration parameter: 'audience'!" unless $conf{'audience'};
	die "Missing required configuration parameter: 'siteName'!" unless $conf{'siteName'};

	# Append "templates" and "public" directories
	my $base = catdir( dirname(__FILE__), 'MozPersona' );
	push @{ $app->renderer->paths }, catdir( $base, 'templates' );
	push @{ $app->static->paths },   catdir( $base, 'public' );

	push @{ $app->renderer->classes }, __PACKAGE__;
	push @{ $app->static->classes },   __PACKAGE__;

	$app->routes->route( $conf{signinPath} )->via('POST')->to(
		namespace         => $conf{namespace},
		action            => 'signin',
		_persona_audience => $conf{audience},
		_persona_service  => $conf{service},
	);
	$app->routes->route( $conf{signoutPath} )->via('POST')->to(
		namespace         => $conf{namespace},
		action            => 'signout',
		_persona_audience => $conf{audience},
		_persona_service  => $conf{service},
	);
	$app->routes->route( $conf{localJsPath} )->via('GET')
		->to( namespace => $conf{namespace}, action => 'js', _persona_conf => \%conf, );

	if ( $conf{'autoHook'} ) {

		my $head_block = '';
		if ( $conf{'autoHook'}->{'css'} ) {
			$head_block
				.= '<link href="/_persona/persona-buttons.css" media="screen" rel="stylesheet" type="text/css" />';
		}
		if ( my $jq = $conf{'autoHook'}->{'jquery'} ) {
			if ( $jq eq 'cdn' ) {
				$head_block
					.= '<script src="http://code.jquery.com/jquery-latest.min.js" type="text/javascript"></script>';
			}
			elsif ( $jq eq 'bundled' or $jq ) {
				$head_block
					.= '<script src="/mojo/jquery/jquery.js" type="text/javascript"></script>';
			}
		}

		my $end_block = '';
		if ( $conf{'autoHook'}->{'persona'} ) {
			$end_block .= qq|<script type="text/javascript" src="$conf{'personaJsPath'}"></script>|;
		}
		if ( $conf{'autoHook'}->{'local'} ) {
			$end_block .= qq|<script type="text/javascript" src="$conf{'localJsPath'}"></script>|;
		}

		$app->hook(
			after_dispatch => sub {
				my ($c) = @_;
				return unless index( $c->res->headers->content_type, 'html' ) >= 0;

				my $body = $c->res->body;

				if ( $conf{'autoHook'}->{'uid'} ) {
					if ( defined( $c->session('_persona') )
						&& $c->session('_persona')->{'status'} eq 'okay' )
					{
						my $email = $c->session('_persona')->{'email'};
						$body
							=~ s!<head>!<head><script type="text/javascript">var personaCurrentUser = "$email";</script>$head_block!o;
					}
					else {
						$body
							=~ s!<head>!<head><script type="text/javascript">var personaCurrentUser = null;</script>$head_block!o;
					}
				}
				elsif ($head_block) {
					$body =~ s!<head>!<head>$head_block!o;
				}

				if ($end_block) {
					$body =~ s!</body>!$end_block</body>!o;
				}
				$c->res->body($body);
			}
		);
	} ## end if ( $conf{'autoHook'})

	return;
} ## end sub register

=head1 HELPERS

None.

=head1 EXTRA FILES

The following files are part of this distribution and are available via
these URL paths:

  /_persona/email_sign_in_black.png
  /_persona/email_sign_in_blue.png
  /_persona/email_sign_in_red.png
  /_persona/persona-buttons.css
  /_persona/persona_sign_in_black.png
  /_persona/persona_sign_in_blue.png
  /_persona/persona_sign_in_red.png
  /_persona/plain_sign_in_black.png
  /_persona/plain_sign_in_blue.png
  /_persona/plain_sign_in_red.png

The images were created by members of the Mozilla Developer Network and are
available at L<Mozilla Persona Branding resources|https://developer.mozilla.org/en-US/docs/persona/branding>.
(c) 2012 by Mozilla Developer Network and / or the individual contributors.

The CSS was created by Sawyer Hollenshead and is also available at
L<Mozilla Persona Branding resources|https://developer.mozilla.org/en-US/docs/persona/branding>.
(c) 2012 by Sawyer Hollenshead.

The following template is part of this distribution:

  _persona/local_js.txt.ep

=head1 SEE ALSO

L<Mojolicious::Plugin::MozPersona::Controller>,
L<https://developer.mozilla.org/en-US/docs/Persona>,
L<Mojolicious>,
L<Mozilla::Persona>.

=cut

1;

__END__

